src/arithmetic/algebraic_expression.c

```diff
@@ -310,93 +209,301 @@ void AlgebraicExpression_PrependTerm(AlgebraicExpression *ae, GrB_Matrix m, bool
     ae->operands[0].operand = m;
 }
 
-AlgebraicExpression **AlgebraicExpression_From_Query(const AST *ast, Vector *matchPattern, const QueryGraph *q, size_t *exp_count) {
-    assert(q->edge_count != 0);
+void AlgebraicExpression_AppendOperand(AlgebraicExpression *ae, AlgebraicExpressionOperand op) {
+    assert(ae);
 
-    AlgebraicExpression *exp = _AE_MUL(q->edge_count + q->node_count);
-    int transpose; // Indicate if matrix operand needs to be transposed.    
-    Node *dest = NULL;
-    Node *src = NULL;
-    Edge *e = NULL;
+    // Make sure expression has enough room for the new operand.
+    if(ae->operand_count+1 > ae->operand_cap) {
+        ae->operand_cap += 4;
+        ae->operands = realloc(ae->operands, sizeof(AlgebraicExpressionOperand) * ae->operand_cap);
+    }
     
-    // Scan MATCH clause from left to right.
-    for(int i = 0; i < Vector_Size(matchPattern); i++) {
-        AST_GraphEntity *match_element;
-        Vector_Get(matchPattern, i, &match_element);
-
-        if(match_element->t != N_LINK) continue;
-        AST_LinkEntity *astEdge = (AST_LinkEntity*)match_element;
-        transpose = (astEdge->direction == N_RIGHT_TO_LEFT);
-        e = QueryGraph_GetEdgeByAlias(q, astEdge->ge.alias);
-        assert(e);
-
-        // TODO: we might want to delay matrix retrieval even further.
-        GrB_Matrix mat = Edge_GetMatrix(e);
-        dest = e->dest;
-        src = e->src;
-
-        if(transpose) {
-            dest = e->src;
-            src = e->dest;
+    ae->operands[ae->operand_count] = op;
+    ae->operand_count++;
+}
+
+void AlgebraicExpression_PrependOperand(AlgebraicExpression *ae, AlgebraicExpressionOperand op) {
+    assert(ae);
+
+    // Make sure expression has enough room for the new operand.
+    ae->operand_count++;
+    if(ae->operand_count+1 > ae->operand_cap) {
+        ae->operand_cap += 4;
+        ae->operands = realloc(ae->operands, sizeof(AlgebraicExpressionOperand) * ae->operand_cap);
+    }
+
+    // TODO: might be optimized with memcpy.
+    // Shift operands to the right, making room at the begining.
+    for(int i = ae->operand_count-1; i > 0 ; i--) {
+        ae->operands[i] = ae->operands[i-1];
+    }
+
+    ae->operands[0] = op;
+}
+
+static void _RemovePathFromGraph(QueryGraph *g, Edge **path) {
+    uint edge_count = array_len(path);
+    for(uint i = 0; i < edge_count; i++) {
+        Edge *e = path[i];
+        Node *src = e->src;
+        Node *dest = e->dest;
+
+        QueryGraph_RemoveEdge(g, e);
+        if(Node_EdgeCount(src) == 0) QueryGraph_RemoveNode(g, src);
+        if(Node_EdgeCount(dest) == 0) QueryGraph_RemoveNode(g, dest);
+    }
+}
+
+/* Determin the length of the longest path in the graph.
+ * Returns a list residing on the edge of the longest path. */
```

Comment:
I didn't quite follow from this comment, so to make sure I understand this function, we're finding the length N of the longest path in the graph, and returning a list of endpoints at one end of all N-length paths?